pymc3/step_methods/hmc/nuts.py

```diff
@@ -239,62 +241,75 @@ def extend(self, direction):
             self.right = tree.right
         else:
             tree, diverging, turning = self._build_subtree(
-                self.left, self.depth, floatX(np.asarray(- self.step_size)))
+                self.left, self.depth, floatX(np.asarray(-self.step_size)))
             self.left = tree.right
 
-        ok = not (diverging or turning)
-        if ok and bern(min(1, tree.size / self.size)):
+        self.depth += 1
+        self.accept_sum += tree.accept_sum
+        self.n_proposals += tree.n_proposals
+
+        if diverging or turning:
+            return diverging, turning
+
+        size1, size2 = self.log_size, tree.log_size
+        if logbern(size2 - size1):
             self.proposal = tree.proposal
 
-        self.depth += 1
-        self.size += tree.size
-        # TODO why not +=
-        #self.accept_sum += tree.accept_sum
-        self.accept_sum = tree.accept_sum
-        #self.n_proposals += tree.n_proposals
-        self.n_proposals = tree.n_proposals
+        self.log_size = np.logaddexp(self.log_size, tree.log_size)
+        self.p_sum[:] += tree.p_sum
 
         left, right = self.left, self.right
-        span = right.q - left.q
-        turning = turning or (span.dot(left.p) < 0) or (span.dot(right.p) < 0)
+        p_sum = self.p_sum
+        turning = (p_sum.dot(left.v) <= 0) or (p_sum.dot(right.v) <= 0)
+
         return diverging, turning
 
     def _build_subtree(self, left, depth, epsilon):
         if depth == 0:
             right = self.leapfrog(left.q, left.p, left.q_grad, epsilon)
             right = Edge(*right)
             energy_change = right.energy - self.start_energy
+            if np.isnan(energy_change):
+                energy_change = np.inf
+
             if np.abs(energy_change) > np.abs(self.max_energy_change):
                 self.max_energy_change = energy_change
             p_accept = min(1, np.exp(-energy_change))
 
-            size = int(self.log_u + energy_change <= 0)
-            diverging = not (self.log_u + energy_change < self.Emax)
+            log_size = -energy_change
+            diverging = energy_change > self.Emax
 
             proposal = Proposal(right.q, right.energy, p_accept)
-            tree = Subtree(right, right, proposal, 1, size, p_accept, 1)
+            tree = Subtree(right, right, right.p, proposal, log_size, p_accept, 1)
             return tree, diverging, False
 
         tree1, diverging, turning = self._build_subtree(left, depth - 1, epsilon)
         if diverging or turning:
             return tree1, diverging, turning
 
         tree2, diverging, turning = self._build_subtree(tree1.right, depth - 1, epsilon)
+        ok = not (diverging or turning)
```

Comment:
this is super descriptive, but might make reading the code easier to just move it to the `if` statement (`if not (diverging or turning):`)